### ISSUE TITLE
ml5 file size

### ISSUE THREAD

**Issue by shiffman** [ml5js/ml5-next-gen#118]
In preparing a ml5.js alpha build, we noticed the filze size is quite large.

<img width="1188" alt="Screen Shot 2024-03-26 at 5 18 43 PM" src="https://github.com/ml5js/ml5-next-gen/assets/191758/5ed2c350-3c17-4c3c-8e97-4d08f3c8d2a5">

It would be worth taking the time to research and look into why this is and if there are dependencies we can remove / scale down. I recall this used to be an issue due to the inclusion of `magenta` but i don't think we are anymore?

**Comment by shiffman** [comment]
So it looks like #111 may have increased the filesize, @ziyuan-linn will take a look.

**Comment by lindapaiste** [comment]
Seems like it's all TensorFlow? Screenshot of the [`webpack-bundle-analyzer`](https://www.npmjs.com/package/webpack-bundle-analyzer).  It does seem like we have both the webgl and cpu backends in the bundle.  We may want to include only 1 backend, so that it works out of the box, and tell the issue to include an additional `<script>` tag to register an other backends.

![image](https://github.com/ml5js/ml5-next-gen/assets/28965286/66084950-89a9-40cf-908f-bd9f0c0a6fcc)

**Comment by ziyuan-linn** [comment]
@lindapaiste Thank you for looking into this. I set `extractComments` to true and it reduced the file size from 4.18 Mib to 3.43 Mib! 

Though I am pretty sure the source map is built to a different file named `ml5.min.js.map`. I tried changing `devtool` to `false` and the bundle size did not change.

**Comment by shiffman** [comment]
This is probably a longer term discussion, but just thinking through some ideas for the future:

- What if we don't include tensorflow.js at all in the bundle? What would be required for ml5.js end users, is it as simple as just including the tf.js script tag as well? Maybe this has no real benefit since the overall size of the webpage wouldn't change.
- The `cpu` back-end is needed for the neuroevolution examples, but is otherwise rarely used. We could consider a build process that creates separate versions `ml5.min.js`, `ml5.webgl.min.js`, `ml5.cpu.min.js`? Or even build ml5.js in a modular way so that beginners can use the entire library, but if someone is building a project that uses only, say, facemesh, they could load `ml5-facemesh.min.js`?

This discussion may become a higher priority if/when we consider bundling other backends for supporting new models, like `transformers.js`.

3.48Mb feels "ok" to me for now so maybe we revisit this after 1.0?

**Comment by MOQN** [comment]
I agree with the idea of organizing the library in a modular way! During a workshop I recently offered at Tongji University in Shanghai, China, I found it challenging to demonstrate examples with multiple iterations because the library's large file size had to be downloaded via a VPN connection.

We can emphasize it on the website and consider adding a friendly error message like: "To use FaceMesh, please import `ml5-facemesh.min.js`. Visit the website for detailed information."

**Comment by shiffman** [comment]
Noting that a discussion here is happening with p5.js which may have some relevant and useful info. cc @sharellb and @gohai as I know you both may be investigating making ml5.js modular.

https://github.com/processing/p5.js/issues/6776

**Comment by quinton-ashley** [comment]
Yes p5.js at ~5MB and p5.min.js at 1.8MB are also quite large.

@sharellb @gohai Just wanted to let y'all know about q5.js, a sequel to p5.js that I'm working on that's only 130kb unminified and it includes WebGPU rendering support!
https://q5js.org/q5.js
